### ISSUE TITLE
Switching to OSGI annotations: How to deal with hyphen ("-") in property ids/names.

### ISSUE THREAD

**Issue by joerghoh** [Adobe-Consulting-Services/acs-aem-commons#1567]
This is related to #1560 : We need to find a way to deal with the fact, that in R6 it's not possible to create ObjectClassDefinition classes which create IDs with a hyphen ("-") in. It's possible in R7. The related chapter in the specification is 105.15.2 (metatype).

**Comment by heervisscher** [comment]
Also with "." we have the issue/challenge

**Comment by royteeuwen** [comment]
> Also with "." we have the issue/challenge

Why? my.property would become String my_property()

**Comment by heervisscher** [comment]
thanks @royteeuwen , didn't know that!

**Comment by joerghoh** [comment]
I see 2 chances to handle property IDs containing a "-":
* Still use SCR annotations for these components/services
* Switch to R7 annotations (only for the annotations!), the runtime should be able to handle it (it already does with the metadata created by the SCR annotations).

These 26 components are using the hypen in a property id at the moment:

 * com.adobe.acs.commons.dam.impl.ReviewTaskAssetMoverHandler.xml (conflict-resolution.default, conflict-resolution.version.last-modified-by, )
 * com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.xml (error-page.extension, error-page.fallback-name, error-page.system-path, not-found.behavior, not-found.exclusion-path-patterns, cache.serve-authenticated, error-images.enabled, error-images.path, error-images.extensions, )
 * com.adobe.acs.commons.hc.impl.HealthCheckStatusEmailer.xml (email.send-only-on-failure, recipients.email-addresses, )
 * com.adobe.acs.commons.http.headers.impl.MonthlyExpiresHeaderFilter.xml (expires.day-of-month, )
 * com.adobe.acs.commons.http.headers.impl.WeeklyExpiresHeaderFilter.xml (expires.day-of-week, )
 * com.adobe.acs.commons.httpcache.config.impl.GroupHttpCacheConfigExtension.xml (httpcache.config.extension.user-groups.allowed, )
 * com.adobe.acs.commons.httpcache.config.impl.HttpCacheConfigImpl.xml (httpcache.config.filter-scope, httpcache.config.cache-handling-rules.pid, )
 * com.adobe.acs.commons.httpcache.config.impl.ResourceTypeHttpCacheConfigExtension.xml (httpcache.config.extension.resource-types.allowed, httpcache.config.extension.resource-types.page-content, )
 * com.adobe.acs.commons.httpcache.engine.impl.HttpCacheEngineImpl.xml (httpcache.engine.cache-handling-rules.global, )
 * com.adobe.acs.commons.images.impl.NamedTransformImageServlet.xml (prop.asset-rendition-picker-regex, )
 * com.adobe.acs.commons.images.transformers.impl.LetterPillarBoxImageTransformerImpl.xml (max-dimension, )
 * com.adobe.acs.commons.images.transformers.impl.ResizeImageTransformerImpl.xml (max-dimension, )
 * com.adobe.acs.commons.oak.impl.EnsureOakIndex.xml (ensure-definitions.path, oak-indexes.path, )
 * com.adobe.acs.commons.replication.dispatcher.impl.DispatcherFlushRulesImpl.xml (prop.replication-action-type, prop.rules.resource-only, )
 * com.adobe.acs.commons.replication.dispatcher.impl.DispatcherFlusherServlet.xml (flush-with-admin-resource-resolver, )
 * com.adobe.acs.commons.replication.status.impl.JcrPackageReplicationStatusEventHandler.xml (node-types, replicated-by.override, replicated-at, )
 * com.adobe.acs.commons.users.impl.EnsureGroup.xml (ensure-immediately, member-of, )
 * com.adobe.acs.commons.users.impl.EnsureServiceUser.xml (ensure-immediately, )
 * com.adobe.acs.commons.util.impl.DelegatingServletFactoryImpl.xml (prop.target-resource-type, )
 * com.adobe.acs.commons.wcm.impl.AemEnvironmentIndicatorFilter.xml (css-color, css-override, inner-html, browser-title-prefix, excluded-wcm-modes, )
 * com.adobe.acs.commons.wcm.impl.CQIncludePropertyNamespaceServlet.xml (namespace.property-names, namespace.property-value-patterns, namespace.multi-level, )
 * com.adobe.acs.commons.wcm.impl.ComponentErrorHandlerImpl.xml (suppress-resource-types, )
 * com.adobe.acs.commons.wcm.views.impl.WCMViewsFilter.xml (path-prefixes.include, resource-types.include, )
 * com.adobe.acs.commons.wcm.views.impl.WCMViewsServlet.xml (wcm-views, )
 * com.adobe.acs.commons.workflow.bulk.removal.impl.WorkflowInstanceRemoverScheduler.xml (workflow.older-than, batch-size, max-duration, )
 * com.adobe.acs.commons.workflow.impl.WorkflowPackageManagerImpl.xml (wf-package.types, )

(created this list with https://gist.github.com/joerghoh/75d440aad3ad0939c5b252be58cd18bb)

**Comment by heervisscher** [comment]
R7 is only available in 6.4

**Comment by heervisscher** [comment]
I tried this with OSGi R7:
	@AttributeDefinition( description = "Configuration values")
	String my$_$property$_$value();

But I got this message in /configMgr: "This form is automatically generated from existing properties because no property descriptors are available for this configuration. This may be cause by the absence of the OSGi Metatype Service or the absence of a MetaType descriptor for this configuration."
AEM6.4.2

**Comment by davidjgonzalez** [comment]
If it’s 6.4 only, then we can’t use R7 annotations since acs commons is compatible w current and previous aem versions. 

I suspect joerg may have found just the property annotations are compatible? (But that’s just a guess?)

**Comment by joerghoh** [comment]
the annotations are used just during compile time, and they are not present during runtime (at least I haven't found an OSGI annotation bundle in AEM yet).

**Comment by davidjgonzalez** [comment]
@heervisscher so does R7 not work in aem 6.4.2 confmgr? (Ie osgi metatype not being generated in a way aem 6.4.2 can understand?)

**Comment by heervisscher** [comment]
It does work, only the thing with the “-“ doesn’t seem to work. For me at least

**Comment by joerghoh** [comment]
I did an experiment over at https://github.com/joerghoh/acs-aem-commons/tree/issue/1631-update-to-r7-annotations, updated the annotations to r7 and changed the R6-property "errorpage.extension" of the ErrorPageHandlerImpl back to its pre-#1626 name "error-page.extension".

```
        String error$_$page_extension();
```
(https://github.com/joerghoh/acs-aem-commons/blob/issue/1631-update-to-r7-annotations/bundle/src/main/java/com/adobe/acs/commons/errorpagehandler/impl/ErrorPageHandlerImpl.java#L113)

I built it and then checked the relevant parts of the `target/classes/OSGI-INF/metatype/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl$Config.xml` (or the 3.19.0 version `target/classes/OSGI-INF/metatype/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.xml`) for the generated metatype statements; I list here the generated statements for the different versions of the annotations

SCR annotations (as of ACS AEM Commons 3.19.0):
```
<metatype:MetaData xmlns:metatype="http://www.osgi.org/xmlns/metatype/v1.0.0" localization="OSGI-INF/metatype/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl">
<AD id="error-page.extension" type="String" default="html" name="%com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.error-page.extension.name" description="%com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.error-page.extension.description"/>
```

osgi r6 annotations (as of Master right now):
```
<metatype:MetaData xmlns:metatype="http://www.osgi.org/xmlns/metatype/v1.2.0" localization="OSGI-INF/l10n/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl$Config">
<AD id="errorpage.extension" type="String" name="Error page extension" description="Examples: html, htm, xml, json. [Optional] [Default: html]" default="html"/>
```

osgi r7 annotations plus the changed property name (my experiment):
```
<metatype:MetaData xmlns:metatype="http://www.osgi.org/xmlns/metatype/v1.4.0" localization="OSGI-INF/l10n/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl$Config">
<AD id="error-page.extension" type="String" name="Error page extension" description="Examples: html, htm, xml, json. [Optional] [Default: html]" default="html"/>
```

When I deploy the r7 version of the ACS AEM Commons bundle to an AEM 6.3 instance I get this exception:
```
9.12.2018 19:12:57.734 *ERROR* [qtp649426978-3968] org.apache.felix.metatype fromDocuments: Error accessing document bundle://525.1:0/OSGI-INF/metatype/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl$Config.xml (java.io.IOException: Unsupported Namespace: 'http://www.osgi.org/xmlns/metatype/v1.4.0')
java.io.IOException: Unsupported Namespace: 'http://www.osgi.org/xmlns/metatype/v1.4.0'
        at org.apache.felix.metatype.MetaDataReader.checkMetatypeNamespace(MetaDataReader.java:209)
        at org.apache.felix.metatype.MetaDataReader.readMetaData(MetaDataReader.java:231)
        at org.apache.felix.metatype.MetaDataReader.parse(MetaDataReader.java:179)
        at org.apache.felix.metatype.MetaDataReader.parse(MetaDataReader.java:119)
        at org.apache.felix.metatype.internal.MetaTypeServiceImpl.fromDocuments(MetaTypeServiceImpl.java:180)
        at org.apache.felix.metatype.internal.MetaTypeServiceImpl.getMetaTypeInformation(MetaTypeServiceImpl.java:136)
        at org.apache.felix.webconsole.internal.configuration.MetaTypeServiceSupport.getObjectClassDefinitions(MetaTypeServiceSupport.java:130)
        at org.apache.felix.webconsole.internal.configuration.MetaTypeServiceSupport.getPidObjectClasses(MetaTypeServiceSupport.java:89)
        at org.apache.felix.webconsole.internal.configuration.ConfigAdminSupport.listConfigurations(ConfigAdminSupport.java:553)
        at org.apache.felix.webconsole.internal.configuration.ConfigManager.renderContent(ConfigManager.java:396)
        at org.apache.felix.webconsole.AbstractWebConsolePlugin.doGet(AbstractWebConsolePlugin.java:194)
        at org.apache.felix.webconsole.internal.configuration.ConfigManager.doGet(ConfigManager.java:331)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:618)
        at javax.servlet.http.HttpServlet.service(HttpServlet.java:725)
```

And this component does not start. Just looking at the `<AD`> tags it should be compatible, but the namespace versions are not. I guess if we just had the chance to revert the namespace definition back to the old 1.2 version, everything would be fine.

**Comment by davidjgonzalez** [comment]
@cziegeler i assume there is no safe way around this? ☝️

**Comment by cziegeler** [comment]
This is done by bnd; i don't know whether it's possible to tell bnd to use a different metatype namespace if possible; but that should be easy to change in bnd
On the other hand you can deploy the latest metatype from Apache Felix; it should not have any other dependencies to be updated

**Comment by joerghoh** [comment]
I want to try another option: Converting the affected components back to the pre-#1626 state and leverage the already contained `org.apache.felix.scrplugin.bnd.SCRDescriptorBndPlugin` to read the SCR annotations. 

This should allow us to use both type of annotations without using the maven-scr-plugin, while still maintaining the compatibility in the metatype names.
After that we can decide if we leave it that way until we switch to R7 completely (then we can rewrite these components to R7 and keep the existing naming) or if we rewrite the code and don't use the "-" in property names any more. 

But it should help us to get the 4.0.0 release out without breaking too much.

**Comment by badvision** [comment]
That SCR plugin is breaking our coveralls testing.  Look @ the logs. 😭

**Comment by badvision** [comment]
I would rather just ditch the hyphens, I don't see why it's so important to keep them.

**Comment by davidjgonzalez** [comment]
because it hard break on a lot of people that require a lot of small changes (room for error) to fix.

**Comment by joerghoh** [comment]
> That SCR plugin is breaking our coveralls testing. Look @ the logs. 😭

you mean the 
```
<_plugin>org.apache.felix.scrplugin.bnd.SCRDescriptorBndPlugin;destdir=${project.build.outputDirectory}</_plugin>
```
statement?

**Comment by joerghoh** [comment]
ok, then my idea does not work. We still could use an oldish version of the maven-scr-plugin ...

**Comment by joerghoh** [comment]
do you have a link to the failures of the jacoco parts? Does it happen as well when you run jacoco locally?

**Comment by badvision** [comment]
mvn -Pcoverage clean prepare-package coveralls:report fails

**Comment by joerghoh** [comment]
Just for validation I reverted the changes of the ErrorPageHandlerImpl again to the 3.19.0 state, added the scr annotations again to the POM and built it (https://github.com/joerghoh/acs-aem-commons/blob/issue/1631-use-scr-again/bundle/src/main/java/com/adobe/acs/commons/errorpagehandler/impl/ErrorPageHandlerImpl.java) . It deploys fine on AEM 6.3 and also the component is starting.

The metatype definition looks like this:
```
<metatype:MetaData xmlns:metatype="http://www.osgi.org/xmlns/metatype/v1.0.0" localization="OSGI-INF/l10n/com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl">
<AD id="error-page.extension" type="String" default="html" name="%com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.error-page.extension.name" description="%com.adobe.acs.commons.errorpagehandler.impl.ErrorPageHandlerImpl.error-page.extension.description"/>
```
As it uses a metatype 1.0 standard, the "-" is just a shortcoming of the official annotations and not of the runtime. And our "only problem" being is the fact, that the annotations write a metatype 1.4 requirements (don't know if we use features of that version at all).

But as @badvision already pointed out, this approach is flawed because now we require the scr-plugin, which breaks jacoco/coveralls.

**Comment by heervisscher** [comment]
Should we just not accept that we can't support '-' in OSGi R6?
Looking at the solutions mentioned above, these are adding IMHO more complexity.

**Comment by justinedelson** [comment]
After playing with a few options, I think I found the one that is the least intrusive.

Basically, whenever there is a an `@AttributeDefinition` which needs to support a backwards compatible name, you add `@LegacyName` to it with the value of that attribute being the legacy name.

Then, in the activate method, you wrap the annotation implementation passed by calling `handleLegacyConfig` passing either the configuration map or a `ComponentContext`.

This internally creates a proxy object which delegates to the original annotation implementation object if no `@LegacyName` is set whereas if `@LegacyName` is set, the value is first attempted to be retrieved from the map and, if not, get the value from the original annotation implementation.

I agree that this is adding more complexity, but I don't think it is significant. ACS AEM Commons has generally done a good job of being backwards compatible and I don't think it should stop now.

This patch isn't complete -- I only annotated two components. But I think it is a good start.

https://github.com/Adobe-Consulting-Services/acs-aem-commons/compare/master...justinedelson:issue/1631-handling-legacy-config-via-proxy?expand=1

h/t to @davidjgonzalez for the idea of using `ValueMap` for this.
